### PR TITLE
drivers: spi: stm32: improve performance with FIFO usage

### PR DESCRIPTION
**Pull request by mniestroj**
Make use of SPI peripheral internal FIFO. Each STM32 family has different
FIFO sizes and some families have per SPI instance FIFO size. Allow to
configure that over devicetree with new `fifo-size` property.

Performance benefits are shown on below testing platforms and Kconfig
options (numbers provided are the duration of nCS pulse being low, in
microseconds):

 * nucleo_l476rg with SPI at 16MHz (CONFIG_SPI_STM32_INTERRUPT=n)

```
   | Transfer direction             |  TX   |  RX   |  TX  |  RX  |
   | Transferred bytes              |  32   |  32   |   4  |   4  |
   |--------------------------------+-------+-------+------+------|
   | CONFIG_SPI_STM32_FIFO=n        |  56.9 |  64.1 |  9.4 | 10.7 |
   | CONFIG_SPI_STM32_FIFO=y        |  36.1 |  37.1 |  9.0 |  9.3 |
```

 * nucleo_l476rg with SPI at 16MHz (CONFIG_SPI_STM32_INTERRUPT=y)

```
   | Transfer direction             |  TX   |  RX   |  TX  |  RX  |
   | Transferred bytes              |  32   |  32   |   4  |   4  |
   |--------------------------------+-------+-------+------+------|
   | CONFIG_SPI_STM32_FIFO=n        | 108.3 | 116.5 | 17.2 | 18.4 |
   | CONFIG_SPI_STM32_FIFO=y        |  39.1 |  39.9 | 12.3 | 12.6 |
```

 * nucleo_h563zi with SPI at 24MHz (CONFIG_SPI_STM32_INTERRUPT=n)

```
   | Transfer direction             |  TX   |  RX   |  TX  |  RX  |
   | Transferred bytes              |  32   |  32   |   4  |   4  |
   |--------------------------------+-------+-------+------+------|
   | CONFIG_SPI_STM32_FIFO=n        |  32.9 |  34.8 |  4.6 |  4.9 |
   | CONFIG_SPI_STM32_FIFO=y        |  18.4 |  18.9 |  3.2 |  4.0 |
```

 * nucleo_h563zi with SPI at 24MHz (CONFIG_SPI_STM32_INTERRUPT=y)

```
   | Transfer direction             |  TX   |  RX   |  TX  |  RX  |
   | Transferred bytes              |  32   |  32   |   4  |   4  |
   |--------------------------------+-------+-------+------+------|
   | CONFIG_SPI_STM32_FIFO=n        |  46.4 |  50.3 |  6.8 |  7.3 |
   | CONFIG_SPI_STM32_FIFO=y        |  19.3 |  20.6 |  3.8 |  5.5 |
```

Tested `tests/drivers/spi/spi_loopback/` with FIFO enabled on following
platforms:
 * `nucleo_g071rb`
 * `nucleo_h563zi`
 * `nucleo_h723zg`
 * `nucleo_l073rz`
 * `nucleo_l476rg`
 * `nucleo_wl55jc`
 * `stm32f0_disco`
 * `stm32f3_disco`

### CODE REVIEW

**Comment by erwango** [comment]
Thanks @mniestroj for this work! @dgastonochoa

**Comment by mniestroj** [comment]
@erwango @dgastonochoa Please let me know what you think on this. I see that a conflicting/alternative solution was already merged few hours before, so I would like some feedback on the approach (just high level) from you both before trying to solve the conflicts. Thanks!

**Comment by erwango** [comment]
Ok, my bad I though this was complementary to @dgastonochoa's work (targeting only H7 like devices).

So, in the general lines, the recommendation on was to make FIFO usage opt-in (thanks to "fifo-enable" prop). Aim is to avoid a situation seen previously in which use of fifo created issues in some use cases that we're not able to reproduce easily in tests.
Note that in merged implementation also allowed to enable it per instance as well.
On  the details of the fifo usage, I think it will require more time to dig in. I'll let @dgastonochoa and @GeorgeCGV have a look as well.

File: drivers/spi/Kconfig.stm32
Comment (dgastonochoa) [inline drivers/spi/Kconfig.stm32:16]: As explained in other comments, enabling/disabling the SPI FIFO is done now per each instance, and from devicetree.

File: drivers/spi/spi_ll_stm32.c
Comment (dgastonochoa) [inline drivers/spi/spi_ll_stm32.c:359]: Writing-then-reading frames in a loop here was discussed [here](https://github.com/zephyrproject-rtos/zephyr/pull/63173#discussion_r1395659932).

I had problems related to the fact that this driver supports different sizes for TX and RX buffers. Here in particular, I think if RX is bigger than TX, the last frames of RX won't be read as TX will have finished and so it won't send dummy zeroes to generate more clock cycles and so be able to read the last RX frames.

I also think I remember having problems related to being inside in the sending loop, then receive some frames to read, then seeing overrun errors because the "send loop" doesn't read. I added a `if (rx_fifo_not_empty()) { read(); break; }` to fix this at some point, but I don't remember the exact circumstances under which the problem appeared. 

Of course, the 2 problems above can be solved by adding more logic, but that adds complexity and I don't see a benefit that justifies so.

Also, why is `tx_fifo_room` required? Why not just write until the fifo is full?

In any case, have you run the SPI loopback tests? I wouldn't merge any change that hasn't passed (at least) those tests.

Could you provide your test plan, please?

File: drivers/spi/spi_ll_stm32.c
Comment (dgastonochoa) [inline drivers/spi/spi_ll_stm32.c:345]: What is the benefit of this change in comparison with the current logic?

**Review by dgastonochoa — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_ll_stm32.c
Comment (tbursztyka) [inline drivers/spi/spi_ll_stm32.c:386]: could it make sense to just call it dfs (aka Data Frame Size) like it is done in many other drivers? (unless ST do not use that term in their specs, I don't know)

**Comment by MaureenHelm** [comment]
@erwango @ajarmouni-st please revisit

**Comment by erwango** [comment]
> @erwango @ajarmouni-st please revisit

Current solution was concurrent (but more complete) to another PR which has been merged in between.
It would be required to be rebased and reshuffled to take into account current driver implementation.

@mniestroj Any plans to do so ,

**Comment by mniestroj** [comment]
Thanks for feedback so far! 

> > @erwango @ajarmouni-st please revisit
> 
> Current solution was concurrent (but more complete) to another PR which has been merged in between. It would be required to be rebased and reshuffled to take into account current driver implementation.
> 
> @mniestroj Any plans to do so ?

Right now I don't have free cycles to work on that, at least in the coming month :( Should we close this PR for now or keep it open?

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Pull request closed without merging**
